pkg/sdk/plugins/source/instance.go

```diff
@@ -0,0 +1,338 @@
+type PushEvent struct {
+	Err       error
+	Data      []byte
+	Timestamp time.Time
+}
+
+type pushInstance struct {
+	builtinInstance
+	evtC <-chan PushEvent
+}
+
+// NewPushInstance opens a new event source and starts a capture session,
+// filling the event batches with a push model.
+//
+// In this model, events are produced through a channel in the form of
+// source.PushEvent messages. This is suitable for cases in which event
+// production is suspensive, meaning that the time elapsed waiting for a
+// new event to be produced is not deterministic or has no guaranteed limit.
+//
+// Users can pass option parameters to influence the behavior of the opened
+// event source, such as passing a context or setting a custom timeout duration.
+//
+// The opened event source can be manually closed by cancelling the optional
+// passed-in context, by closing the event cannel, or by sending
+// source.PushEvent containing a non-nil Err.
+func NewPushInstance(evtC <-chan PushEvent, options ...func(*builtinInstance)) (Instance, error) {
+	res := &pushInstance{
+		evtC: evtC,
+		builtinInstance: builtinInstance{
+			ctx:       context.Background(),
+			timeout:   defaultInstanceTimeout,
+			shutdown:  func() {},
+			eof:       false,
+			batchSize: sdk.DefaultBatchSize,
+			eventSize: sdk.DefaultEvtSize,
+		},
+	}
+
+	// apply options
+	for _, opt := range options {
+		opt(&res.builtinInstance)
+	}
+
+	// create custom-sized event batch
+	batch, err := sdk.NewEventWriters(int64(res.batchSize), int64(res.eventSize))
+	if err != nil {
+		return nil, err
+	}
+	res.SetEvents(batch)
+
+	// init timer
+	res.timeoutTicker = time.NewTicker(res.timeout)
+
+	// setup internally-cancellable context
+	prevCancel := res.shutdown
+	cancelableCtx, cancelCtx := context.WithCancel(res.ctx)
+	res.ctx = cancelableCtx
+	res.shutdown = func() {
+		cancelCtx()
+		prevCancel()
+	}
+
+	return res, nil
+}
+
+func (s *pushInstance) NextBatch(pState sdk.PluginState, evts sdk.EventWriters) (int, error) {
+	// once EOF has been hit, we should return it at each new call of NextBatch
+	if s.eof {
+		return 0, sdk.ErrEOF
+	}
+
+	// timeout needs to be resetted for this batch
+	s.timeoutTicker.Reset(s.timeout)
+
+	// attempt filling the event batch
+	n := 0
+	for n < evts.Len() {
+		select {
+		// an event is received, so we add it in the batch
+		case evt, ok := <-s.evtC:
+			// event channel is closed, we reached EOF
+			if !ok {
+				evt.Err = sdk.ErrEOF
+			}
+			// if there are no errors so far, try writing the event
+			if evt.Err == nil {
+				if l, wErr := evts.Get(n).Writer().Write(evt.Data); wErr != nil {
+					evt.Err = wErr
+				} else if l < len(evt.Data) {
+					evt.Err = io.ErrShortWrite
+				}
+			}
+			// an error occurred, so we need to exit
+			if evt.Err != nil {
+				s.eof = true
+				return n, evt.Err
+			}
+			// event added to the batch successfully
+			if !evt.Timestamp.IsZero() {
+				evts.Get(n).SetTimestamp(uint64(evt.Timestamp.UnixNano()))
+			}
```

Comment:
```suggestion
			if evt.Timestamp.IsZero() {
				evts.Get(n).SetTimestamp(math.MaxUint64)
			} else {
				evts.Get(n).SetTimestamp(uint64(evt.Timestamp.UnixNano()))
			}
```

Since `evts` are being reused, if a timestamp is set previously and then the timestamp is zero to the next batch iteration it won't be reset, right?